### PR TITLE
LLM-Assisted Translations Frontend (rebased)

### PR DESCRIPTION
**Pull request by mathjazz**
This is a continuation of https://github.com/mozilla/pontoon/pull/3127, which got stuck due to a bad rebase.

I've rebased it against main and pushed to a new branch to be able to roll back. For all conflicts, I just accepted incoming changes and added [40bf35e](https://github.com/mozilla/pontoon/pull/3159/commits/40bf35e3c94f43ccf3e330661fad75911571f1f5).

@ayanaar Could you please check that this works as expected and contains all and only relevant changes? If yes, we should close the original PR.

### CODE REVIEW

**Review by mathjazz — COMMENTED** [review]
Just a few nits for now.

File: translate/src/api/machinery.ts [hunk 1]
```diff
@@ -140,6 +141,45 @@ export async function fetchGoogleTranslation(
     : [];
 }
 
+/**
+ * Return refined translation by GPT-4.
+ */
+
+export async function fetchGPTTransformation(
```
[the remark this answers is not in the document]
Comment (mathjazz) [reply translate/src/api/machinery.ts:148]: Transform sounds more consistent with the rest of the related names.

File: translate/src/modules/machinery/components/MachineryTranslationSource.tsx [hunk 2]
```diff
@@ -1,5 +1,3 @@
-import React from 'react';
-
```
[the remark this answers is not in the document]
Comment (ayanaar) [reply translate/src/modules/machinery/components/MachineryTranslationSource.tsx]: It seems to be working on my end. I never got any errors for some reason. I will definitely add it back though.

File: translate/src/modules/machinery/components/source/GoogleTranslation.tsx [hunk 3]
```diff
@@ -1,27 +1,156 @@
...
+            machineryTranslations[0].translation.replace(
+              /^['"](.*)['"]$/,
+              '$1',
+            );
+          setCurrentTranslation(translationWithoutQuotes);
+          setLLMTranslation(translationWithoutQuotes);
+          setShowOriginalOption(true);
+        }
+      } else {
+        console.log(translation.translation);
```
Comment (eemeli) [inline translate/src/modules/machinery/components/source/GoogleTranslation.tsx:67]:
```suggestion
```

File: translate/src/modules/machinery/components/source/GoogleTranslation.tsx [hunk 4]
```diff
@@ -1,27 +1,156 @@
...
+          characteristic,
+          locale.name,
+        );
+
+        if (machineryTranslations.length > 0) {
+          const translationWithoutQuotes =
+            machineryTranslations[0].translation.replace(
+              /^['"](.*)['"]$/,
+              '$1',
+            );
```
Comment (eemeli) [inline translate/src/modules/machinery/components/source/GoogleTranslation.tsx:61]: This sort of cleanup belongs in `fetchGPTTtransform()`, not here.

File: translate/src/modules/machinery/components/source/GoogleTranslation.tsx [hunk 5]
```diff
@@ -1,27 +1,156 @@
...
+      } else {
+        console.log(translation.translation);
+        setCurrentTranslation(translation.translation);
+        setLLMTranslation('');
+        setSelectedOption('');
+        setShowOriginalOption(false);
+      }
+    } catch (error) {
+      console.error('Error fetching GPT transformation:', error);
+    }
```
Comment (eemeli) [inline translate/src/modules/machinery/components/source/GoogleTranslation.tsx:75]: There's already a try-catch in `fetchGPTTtransform()` to swallow errors, anything that might throw should be done there, and we should not need one here.

File: translate/src/modules/machinery/components/source/GoogleTranslation.tsx [hunk 6]
```diff
@@ -1,27 +1,156 @@
...
+    });
+
+    handleTransformation(ev, characteristic);
+  };
+
   return (
-    <li>
+    <li ref={dropdownRef} className='google-translation'>
       <Localized
         id='machinery-GoogleTranslation--visit-google'
```
Comment (eemeli) [inline translate/src/modules/machinery/components/source/GoogleTranslation.tsx:118]: This class needs to be renamed if this isn't a link.

File: translate/src/modules/machinery/components/source/GoogleTranslation.tsx [hunk 7]
```diff
@@ -1,27 +1,156 @@
...
+  const [showOriginalOption, setShowOriginalOption] = useState(false);
+  const [currentTranslation, setCurrentTranslation] = useState(
+    translation.translation,
+  );
+  const dropdownRef = useRef<HTMLLIElement>(null);
+  const locale = useContext(Locale);
+  const [llmTranslation, setLLMTranslation] = useState('');
+
+  const toggleDropdown = (ev: React.MouseEvent) => {
+    ev.stopPropagation();
```
Comment (eemeli) [inline translate/src/modules/machinery/components/source/GoogleTranslation.tsx:32]: Why is this line needed? Same applies to all `stopPropagation()` calls below, I'm pretty sure none of them have any effect.

File: translate/src/modules/machinery/components/source/GoogleTranslation.tsx [hunk 8]
```diff
@@ -1,27 +1,156 @@
...
+  useEffect(() => {
+    // Whenever llmTranslation changes, communicate this change to the parent component
+    onLLMTranslationChange(llmTranslation);
+  }, [llmTranslation]);
+
+  const handleTransformation = async (
+    ev: React.MouseEvent,
+    characteristic: string,
+  ) => {
+    ev.stopPropagation();
```
Comment (eemeli) [inline translate/src/modules/machinery/components/source/GoogleTranslation.tsx:45]: This does not need the event as an argument.

File: translate/src/modules/machinery/components/source/GoogleTranslation.tsx [hunk 9]
```diff
@@ -1,27 +1,156 @@
...
+        onClick={toggleDropdown}
+        className='dropdown-toggle'
+        aria-haspopup='true'
+        aria-expanded={isDropdownOpen}
+      >
+        <i className='fa fa-caret-down'></i>
+      </button>
+      {isDropdownOpen && (
+        <ul className='dropdown-menu' style={{ display: 'block' }}>
+          <li onClick={(ev) => handleOptionClick(ev, 'Rephrase')}>REPHRASE</li>
```
Comment (eemeli) [inline translate/src/modules/machinery/components/source/GoogleTranslation.tsx:140]: Should refactor the callback so something like this works:

```suggestion
          <li data-characteristic="alternative" onClick={handleOptionClick}>Rephrase</li>
```

Also, we could and should do the capitalization with CSS.

**Review by eemeli — CHANGES_REQUESTED** [review]

File: pontoon/base/migrations/0057_alter_translation_machinery_sources.py [hunk 10]
```diff
@@ -0,0 +1,36 @@
+# Generated by Django 3.2.24 on 2024-04-02 01:36
+
+import django.contrib.postgres.fields
+from django.db import migrations, models
+
+
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ("base", "0056_remove_fuzzy_and_non-approved_tm_entries"),
```
[the remark this answers is not in the document]
Comment (mathjazz) [reply pontoon/base/migrations/0057_alter_translation_machinery_sources.py:10]: Yeah, we need to rebase again.

File: translate/src/modules/machinery/components/MachineryTranslation.css [hunk 11]
```diff
@@ -44,6 +44,17 @@
   padding-right: 3px;
 }
 
+.machinery
+  .translation
+  > header
+  ul
+  li
+  .dropdown-menu
+  li:not(:first-child)::before {
```
[the remark this answers is not in the document]
Comment (mathjazz) [reply translate/src/modules/machinery/components/MachineryTranslation.css:53]: @ayanaar As agreed, I'll make some UI changes on top of this PR, so I can take care of the CSS review comments as part of that.

File: pontoon/base/migrations/0057_alter_translation_machinery_sources.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (eemeli) [reply pontoon/base/migrations/0057_alter_translation_machinery_sources.py:10]: Note that now #3169 is adding a `0058` as well.

File: translate/src/modules/machinery/components/source/GoogleTranslation.tsx (hunk 7 above)
Comment (ayanaar) [reply translate/src/modules/machinery/components/source/GoogleTranslation.tsx:32]: This is needed as well as the call in the `handleOptionClick` function. Without these calls, clicking on the dropdown menu causes the text in the machinery tab to be copied into the editor (and the dropdown menu does not open).

**Comment by eemeli** [comment]
With the rebase I can't tell what changes were made after my last review. What's the diff or commit range for that? The automatic one GitHub provides above includes the .lang removal and a bunch of other stuff.

For next time, doing a merge from main to this branch can provide the same results, but with a history that's much easier for a reviewer to navigate. We'll be squashing changes in any case, so feature branches don't need to stay linear.

**Comment by ayanaar** [comment]
> > What's the diff or commit range for that?
> 
> Last 4 commits are new.

The commit labelled [Refactoring based on review](https://github.com/mozilla/pontoon/pull/3159/commits/ec8f0d83cd8d4ed015c5955650866d24b1b57415) has most of the requested changes from your review. The following commits I just added a migration file and made the dropdown menu content localizable.

File: pontoon/machinery/openai_service.py [hunk 12]
```diff
@@ -20,29 +21,32 @@ def get_translation(
                 f"The target language '{target_language_name}' is not supported."
             )
 
-        informal = (
-            f"You will be provided with text in English, along with its machine-generated translation in {target_language}. "
+        informal = textwrap.dedent(
+            f"""
+            You will be provided with text in English, along with its machine-generated translation in {target_language}. "
```
Comment (eemeli) [inline pontoon/machinery/openai_service.py:26]: The trailing quote here and the quotes on the following lines are now a part of the string value.

File: translate/src/api/machinery.ts [hunk 13]
```diff
@@ -164,12 +164,13 @@ export async function fetchGPTTransformation(
     const { translation } = (await GET_(url, params)) as {
       translation: string;
     };
+    const cleanedTranslation = translation.replace(/^['"](.*)['"]$/, '$1');
     if (translation) {
```
Comment (eemeli) [inline translate/src/api/machinery.ts:168]:
```suggestion
    if (translation) {
      const cleanedTranslation = translation.replace(/^['"](.*)['"]$/, '$1');
```

File: translate/src/modules/machinery/components/source/GoogleTranslation.tsx [hunk 14]
```diff
@@ -38,82 +38,70 @@ export function GoogleTranslation({
...
     }
   };
 
-  const handleOptionClick = (ev: React.MouseEvent, option: string) => {
+  const handleOptionClick = (ev: React.MouseEvent<HTMLLIElement>) => {
     ev.stopPropagation();
-    setSelectedOption(option);
-    setDropdownOpen(false);
+    const target = ev.currentTarget;
+    const characteristic = target.getAttribute('data-characteristic');
```
Comment (eemeli) [inline translate/src/modules/machinery/components/source/GoogleTranslation.tsx:67]:
```suggestion
    const characteristic = target.dataset['characteristic'];
```

**Review by eemeli — CHANGES_REQUESTED** [review]
Will do a proper review tomorrow, but noticed a couple of small things.

File: pontoon/machinery/openai_service.py [hunk 15]
```diff
@@ -20,29 +21,32 @@ def get_translation(
...
             )
 
-        informal = (
-            f"You will be provided with text in English, along with its machine-generated translation in {target_language}. "
-            "Your objective is to revise the {target_language} translation to ensure it utilizes simpler language. Adhere to the following guidelines to achieve this:\n"
-            "- Clarity is Key: Ensure the translation conveys the original message in the clearest possible manner, without ambiguity or unnecessary complexity.\n"
-            "- Consistent Simplicity: Maintain a consistent level of simplicity throughout the translation.\n"
-            "The goal is to produce a translation that accurately reflects the original English text, but in a way that is more approachable and easier to understand for all {target_language} speakers."
+        informal = textwrap.dedent(
+            f"""
```
Comment (eemeli) [inline pontoon/machinery/openai_service.py:25]: Should drop the leading newline.
```suggestion
            f"""\
```

File: pontoon/machinery/openai_service.py [hunk 16]
```diff
@@ -20,29 +21,32 @@ def get_translation(
...
 
-        formal = (
-            f"You will be provided with text in English, along with its machine-generated translation in {target_language}. "
-            "Your objective is to revise the {target_language} translation to ensure it utilizes a higher level of formality. Adhere to the following guidelines to achieve this:\n"
-            "- Adjust the Tone: Ensure the tone is respectful, polished, and devoid of colloquialisms or informal expressions commonly used in casual conversation.\n"
-            "- Formal Addressing: Where applicable, use formal modes of address.\n"
-            "- Consistency: Maintain a consistent level of formality throughout the translation, avoiding shifts in tone or style.\n"
-            "Your goal is to produce a translation that not only accurately conveys the meaning of the English text but also meets the expectations for formality in {target_language}-speaking professional or formal settings."
+        formal = textwrap.dedent(
+            f"""
```
Comment (eemeli) [inline pontoon/machinery/openai_service.py:34]:
```suggestion
            f"""\
```

File: pontoon/machinery/openai_service.py [hunk 17]
```diff
@@ -20,29 +21,32 @@ def get_translation(
...
         )
 
-        alternative = (
-            f"You will be provided with text in English, along with its machine-generated translation in {target_language}. "
-            "Your objective is to provide an alternative translation. Adhere to the following guidelines to achieve this:\n"
-            "- Cultural Nuances: Pay attention to cultural nuances and idiomatic expressions, ensuring they are appropriately translated for the {target_language}-speaking audience.\n"
-            "- Clarification and Accuracy: Where the source text is ambiguous or idiomatic, offer clarifications or alternative expressions in {target_language}.\n"
-            "Just provide the text for the alternative translation."
+        alternative = textwrap.dedent(
+            f"""
```
Comment (eemeli) [inline pontoon/machinery/openai_service.py:44]:
```suggestion
            f"""\
```

Comment (eemeli) [inline pontoon/machinery/openai_service.py:44]:
```suggestion
            f"""\
```

File: translate/src/modules/machinery/components/source/GoogleTranslation.tsx (hunk 6 above)
Comment (eemeli) [reply translate/src/modules/machinery/components/source/GoogleTranslation.tsx:106]: Ach, sorry, it looks like I mis-read this earlier. I wasn't commenting on a class here, but an L10n identifier, and that's still incorrectly applying a title "Visit Google Translate" to this element.

File: translate/src/modules/machinery/components/source/GoogleTranslation.tsx [hunk 18]
```diff
@@ -38,82 +38,70 @@ export function GoogleTranslation({
...
+        case 'informal':
+          displayText = 'INFORMAL';
+          break;
+        case 'original':
+          displayText = '';
+          break;
+        default:
+          break;
+      }
+      setSelectedOption(displayText);
```
Comment (eemeli) [inline translate/src/modules/machinery/components/source/GoogleTranslation.tsx:88]: If the dropdown options are localised, the selected value display should also be localised. But this is messy enough already, so I'd be fine with a "TODO" or "FIXME" code comment here to that effect.

File: translate/src/modules/machinery/components/source/GoogleTranslation.tsx [hunk 19]
```diff
@@ -38,82 +38,70 @@ export function GoogleTranslation({
...
+      setDropdownOpen(false);
 
-    logUXAction('LLM Dropdown Select', 'LLM Feature Adoption', {
-      optionSelected: option,
-      targetLanguage: locale.name,
-    });
+      const trackllmTranslation = target.innerText;
+      onLLMTranslationChange(trackllmTranslation);
```
Comment (eemeli) [inline translate/src/modules/machinery/components/source/GoogleTranslation.tsx:92]: Is this right? As far as I can tell, `target.innerText` here is the localized `<li>` contents, so something like "MAKE INFORMAL", which is now ending up as the `llmtranslation` used in `copyLLMTranslationIntoEditor` and from there via keyboard shortcut in the editor.

File: translate/src/modules/machinery/components/source/GoogleTranslation.tsx [hunk 20]
```diff
@@ -38,82 +38,70 @@ export function GoogleTranslation({
...
 
-    handleTransformation(ev, characteristic);
+      logUXAction('LLM Dropdown Select', 'LLM Feature Adoption', {
+        optionSelected: trackllmTranslation,
```
Comment (eemeli) [inline translate/src/modules/machinery/components/source/GoogleTranslation.tsx:95]: Wouldn't this be better than the localized text?

```suggestion
        optionSelected: characteristic,
```

**Review by eemeli — CHANGES_REQUESTED** [review]

File: translate/src/modules/machinery/components/source/GoogleTranslation.tsx (hunk 19 above)
Comment (ayanaar) [reply translate/src/modules/machinery/components/source/GoogleTranslation.tsx]: I was using the text as a placeholder for the loader (i.e, "MAKE INFORMAL" would show up in until the LLM result loaded) but I have removed it now.

File: translate/public/locale/en-US/translate.ftl [hunk 21]
```diff
@@ -503,7 +503,7 @@ machinery-ConcordanceSearch--translation-memory = TRANSLATION MEMORY
 
 ## Machinery Google Translation
 ## Shows the translation source from Google Translate.
-machinery-GoogleTranslation--visit-google = GOOGLE TRANSLATE
+machinery-GoogleTranslation--translation-source = GOOGLE TRANSLATE
     .title = Visit Google Translate
```
Comment (eemeli) [inline translate/public/locale/en-US/translate.ftl:507]: This is the bit that's incorrect now that it's no longer a link.
```suggestion
```

File: translate/src/modules/machinery/components/source/GoogleTranslation.tsx [hunk 22]
```diff
@@ -1,27 +1,151 @@
...
+}: Props): React.ReactElement<'li'> {
+  const [isDropdownOpen, setDropdownOpen] = useState(false);
+  const [selectedOption, setSelectedOption] = useState('');
+  const [showOriginalOption, setShowOriginalOption] = useState(false);
+  const [currentTranslation, setCurrentTranslation] = useState(
+    translation.translation,
+  );
+  const dropdownRef = useRef<HTMLLIElement>(null);
+  const locale = useContext(Locale);
+  const [llmTranslation, setLLMTranslation] = useState('');
```
Comment (eemeli) [inline translate/src/modules/machinery/components/source/GoogleTranslation.tsx:29]: The passing of the `llmTranslation` value up the stack is a bit of a code smell that there really ought to be a Context handling it, but at least it should be possible to avoid the `useEffect()` with a wrapper like this:

```suggestion
  const [llmTranslation, _setLLMTranslation] = useState('');
  const setLLMTranslation = (newTranslation: string) => {
    _setLLMTranslation(newTranslation);
    onLLMTranslationChange(newTranslation);
  }
```

File: translate/src/modules/machinery/components/source/GoogleTranslation.tsx [hunk 23]
```diff
@@ -1,27 +1,151 @@
...
+  const toggleDropdown = (ev: React.MouseEvent) => {
+    ev.stopPropagation();
+    setDropdownOpen((isDropdownOpen) => !isDropdownOpen);
+  };
+
+  useEffect(() => {
+    // Whenever llmTranslation changes, communicate this change to the parent component
+    onLLMTranslationChange(llmTranslation);
+  }, [llmTranslation]);
+
```
Comment (eemeli) [inline translate/src/modules/machinery/components/source/GoogleTranslation.tsx:40]:
```suggestion
```

File: translate/src/modules/machinery/components/source/GoogleTranslation.tsx [hunk 24]
```diff
@@ -1,27 +1,151 @@
-import React from 'react';
+import React, { useState, useRef, useContext, useEffect } from 'react';
```
Comment (eemeli) [inline translate/src/modules/machinery/components/source/GoogleTranslation.tsx:1]:
```suggestion
import React, { useState, useRef, useContext } from 'react';
```

**Review by eemeli — CHANGES_REQUESTED** [review]
Nearly there! A couple of small fixes still inline.

File: translate/public/locale/en-US/translate.ftl (hunk 21 above)
Comment (ayanaar) [reply translate/public/locale/en-US/translate.ftl:507]: As discussed in our call, we are keeping this for now.

**Review by eemeli — APPROVED** [review]

**Comment by mathjazz** [comment]
I've landed CSS enhancements in https://github.com/mozilla/pontoon/pull/3159/commits/f4efd51a656e7bb98daba6fbca342cb12f26c932. @ayanaar Could you please take a look?

I also found a bug. If you click outside the `.translation-source` element, text in `.selected-option` disappears and the menu doesn't open. Please fix that.

Also, could you please create a placeholder for the loading animation? No need to do anything with the design, just make sure that something like `<span>Loading...</span>` appears after the selection is made and before the actual LLM translation is shown.

File: translate/src/context/TranslationContext.tsx [hunk 25]
```diff
@@ -1,100 +1,84 @@
...
-        displayText = 'FORMAL';
-        break;
-      case 'informal':
-        displayText = 'INFORMAL';
-        break;
-      case 'original':
-        displayText = '';
-        break;
-    }
+  useEffect(() => setState(new Map()), [translation]);
```
Comment (eemeli) [inline translate/src/context/TranslationContext.tsx:35]: This is here to prevent a memory leak.

File: translate/src/context/TranslationContext.tsx [hunk 26]
```diff
@@ -1,100 +1,84 @@
...
-      case 'formal':
-        displayText = 'FORMAL';
-        break;
-      case 'informal':
-        displayText = 'INFORMAL';
-        break;
-      case 'original':
-        displayText = '';
-        break;
-    }
```
Comment (eemeli) [inline translate/src/context/TranslationContext.tsx]: This is more appropriate to be in the component, where it'll also be easier to refactor later for proper localization. I think there was a code comment on that which has been dropped?

File: translate/src/context/TranslationContext.tsx [hunk 27]
```diff
@@ -1,100 +1,84 @@
...
+    mt: MachineryTranslation,
     characteristic: string,
     localeName: string,
   ) => {
-    setLoading(true);
+    let { llmTranslation, selectedOption } = getSelState(mt);
     if (characteristic !== 'original') {
+      let next = new Map(state);
+      next.set(mt, { loading: true, selectedOption, llmTranslation });
+      setState(next);
```
Comment (eemeli) [inline translate/src/context/TranslationContext.tsx:49]: Repeating this three times is just on the edge of maaaaybe wanting to get wrapped in a function.

File: translate/src/context/TranslationContext.tsx [hunk 28]
```diff
@@ -1,100 +1,84 @@
...
 
 export const LLMTranslationProvider: React.FC = ({ children }) => {
-  const [llmTranslation, setLlmTranslation] = useState<string>('');
-  const [loading, setLoading] = useState<boolean>(false);
-  const [selectedOption, setSelectedOption] = useState<string>('');
-
-  const handleSetSelectedOption = (option: string) => {
-    let displayText = '';
+  const translation = useActiveTranslation();
+  const [state, setState] = useState(new Map<MachineryTranslation, SelState>());
```
[the remark this answers is not in the document]
Comment (ayanaar) [reply translate/src/context/TranslationContext.tsx]: Thanks for the explanation. Based on your suggestion I have implemented a WeakMap for state management to efficiently handle multiple translation states without retaining unnecessary objects, and added a version counter to ensure the UI updates correctly.

File: translate/src/context/TranslationContext.tsx (hunk 26 above)
Comment (ayanaar) [reply translate/src/context/TranslationContext.tsx]: Thanks for catching that, I have added the comment back.

File: translate/src/context/TranslationContext.tsx [hunk 29]
```diff
@@ -29,42 +34,49 @@ const LLMTranslationContext = createContext<LLMTranslationContextType>({
...
-      }
-      next = new Map(state);
-      next.set(mt, { loading: false, selectedOption, llmTranslation });
-      setState(next);
+    const currentState = getSelState(mt);
+    stateRef.current.set(mt, {
+      ...currentState,
+      loading: true,
+      selectedOption: currentState.selectedOption,
+      llmTranslation: currentState.llmTranslation,
```
Comment (eemeli) [inline translate/src/context/TranslationContext.tsx:57]: These lines don't do anything, given the `...currentState` above.
```suggestion
```

File: translate/src/context/TranslationContext.tsx [hunk 30]
```diff
@@ -29,42 +34,49 @@ const LLMTranslationContext = createContext<LLMTranslationContextType>({
...
-  const translation = useActiveTranslation();
-  const [state, setState] = useState(new Map<MachineryTranslation, SelState>());
+  const stateRef = useRef(new WeakMap<MachineryTranslation, SelState>());
+  const [version, setVersion] = useState(0); // Counter to trigger re-renders
 
-  useEffect(() => setState(new Map()), [translation]);
-
-  const getSelState = (mt: MachineryTranslation) =>
-    state.get(mt) ?? initSelState();
+  const getSelState = useCallback(
```
Comment (eemeli) [inline translate/src/context/TranslationContext.tsx:40]: Wrapping this in a callback is unnecessary, as the only time this component re-renders is when `version` changes.

File: translate/src/context/TranslationContext.tsx [hunk 31]
```diff
@@ -35,14 +29,11 @@ const LLMTranslationContext = createContext<LLMTranslationContextType>({
 
 export const LLMTranslationProvider: React.FC = ({ children }) => {
   const stateRef = useRef(new WeakMap<MachineryTranslation, SelState>());
-  const [version, setVersion] = useState(0); // Counter to trigger re-renders
+  const [_, setVersion] = useState(0); // Counter to trigger re-renders
```
Comment (eemeli) [inline translate/src/context/TranslationContext.tsx:32]: The unused variable can be elided completely.
```suggestion
  const [, setVersion] = useState(0); // Counter to trigger re-renders
```

**Review by eemeli — APPROVED** [review]

**Comment by exsilium** [comment]
Hi, Just wanted to congratulate people working on the LLM feature and great work on this PR. I've been following the progress closely and I'm really looking forward for this feature to land. I appreciate the dedication and effort that has gone into it. Can't wait for the merge to happen (once conflicts are finally resolved 😄 ) and start experimenting with it. Thank you for all the hard work with this! Cheers! 🍻 👍 💯 🥇

**Comment by mathjazz** [comment]
Closing on behlaf of https://github.com/mozilla/pontoon/pull/3191.
